Ginkgo Test: Add benchmark class of tests and an initial netperf test

jrfastab opened the pull request:
This adds a super_netperf test that reports metrics in a l4bench_perf.log file for an increasing number of instances. We do two test configures initially (a) netperf over loopback and (b) netperf from endpoint to endpoint. The test is run twice first without sockmap enabled and then again with it enabled.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6193)
<!-- Reviewable:end -->

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+package benchmarktest
```
Add copyright header.

jrfastab replied:
thanks done

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+package benchmarktest
```
Typo in commit - allowsjus --> allows us

jrfastab replied:
yep thanks.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+		helpers.ExpectCiliumReady(vm)
+	})
+
+	AfterAll(func() {
```
Remove this if it's empty.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	removeContainers := func(containerName string) {
+		By(fmt.Sprintf("removing container %s", containerName))
```
You can remove the `fmt.Sprintf` - formatting is done within `By`.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+	}, 500)
+
+	createContainers := func() {
+		vm.ContainerCreate(helpers.Client, helpers.NetperfImage, helpers.CiliumDockerNetwork, "-l id.client")
```
Add a `By` indicating that you are creating the containers in this function.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+			}
+			return true
+		}, "Endpoints are not ready", &helpers.TimeoutConfig{Timeout: 150})
+		Expect(err).Should(BeNil())
```
Add a string as a second parameter to `Should`, e.g. `Should(BeNil(), fmt.Sprintf("timed out waiting for endpoints waiting to be ready while creating containers")`

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+			res.ExpectSuccess()
+		}
+
+		SuperNetperfRRIPv4 := func(client bool, num int, serverIP string) {
```
Can you add comments indicating what these parameters do? Will make it easier for future editors of this code to grasp what it does.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+
+			By("checking %s can ping to %s IPv4", helpers.Client, helpers.Server)
+			res = vm.ContainerExec(helpers.Client, helpers.Ping(serverIP))
+			res.ExpectSuccess()
```
Add a string parameter to this indicating what happened, e.g. "error pinging server from client". Do this for each `ExpectSuccess` invocation.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+			ClientServerCheck(serverIPv6, serverIP)
+		}
+
+		NetperfTest := func() {
```
I'd recommend splitting up this entire test into separate `Its`, where each test runs a `SuperNetPerf` test in each `It`. This makes it so that each `It` corresponds to roughly one assertion as opposed to running everything within one `It`. You could get the `serverData` and `serverIPv6` information outside of the `It` (say in `BeforeAll`), and then have each `It` run the netperf tests.

Example:

```
It("L4 Netperf over loopback with 1 connection", func() {
    SuperNetPerfRR(false, 1, serverIP, serverIPv6)
})
```

@eloycoto please confirm what I am saying above, as I believe you have vocalized opinions that echo what I have stated before.

jrfastab replied:
Will the test_results get a bit noisy if we do this? I think we get a directory per It instance.

eloycoto replied:
@jrfastab do as Ian mentionend. The test results does not matter at all because the output will be  attached in the Jenkins so it's easy   ;-)

jrfastab replied:
OK will do this today.

ianvernon commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+			NetperfTest()
+		}, 300)
+
+		It("Test L4 Netperf Performance with Sockmap", func() {
```
To do what I stated above by separating the tests into individual `It`, you could put this in a separate context called `SockMap Netperf`

ianvernon commented on test/test_test.go:
```diff
@@ -16,6 +16,7 @@ package ciliumTest
+	_ "github.com/cilium/cilium/test/benchmark"
 	_ "github.com/cilium/cilium/test/k8sT"
 	_ "github.com/cilium/cilium/test/runtime"
 )
```
There may need to be more work done to handle this new `Benchmark` scope - cc: @eloycoto .

joestringer replied:
I think that the scope shouldn't need updating unless we create a new topology. Spoke offline with @jrfastab about this and he's been trying this with Runtime scope.

joestringer commented on test/benchmark/l4bench.go:
```diff
@@ -0,0 +1,194 @@
+		ClientServerCheck := func(serverIPv6 string, serverIP string) {
+			By("checking %s can ping to %s IPv6", helpers.Client, helpers.Server)
+			res := vm.ContainerExec(helpers.Client, helpers.Ping6(serverIPv6))
+			res.ExpectSuccess()
```
When this `ExpectSuccess()` fails, the failure in the ginkgo output will print that this line is the point of failure and it might not be obvious where the failure actually occurred from the test context (eg the fact that it was the 5th `SuperNetperfRR` call).

We will often work around this by doing something like:

```
ExpectWithOffset(1, res).To(helpers.CMDSuccess(), "Connectivity check failed")
```

Where 1 is the distance from this function to the actual test function (`Describe()`, `It()`, ...) that is calling this helper.

Same feedback applies to the rest of this helper and the other helper functions you've defined in this Context.

jrfastab replied:
Now that everything is broken into It() branches and netperf code pulled into helpers/benchmark I think we can skip ExpectWithOffset and it will still be clear what failed and where.

joestringer reviewed (commented):
Just one comment about helper functions and the offset that ginkgo will report in failure conditions.

eloycoto commented on test/helpers/assertions.go:
```diff
@@ -0,0 +1,31 @@
+//
```
This should be in runtime/assertions.go

eloycoto commented on test/helpers/benchmark.go:
```diff
@@ -0,0 +1,71 @@
+
+// ClientServerCheck verifies client can still ping server with
+// both IPv4 and IPv6
+func (s *SSHMeta) ClientServerCheck(client string, server string) {
```
Why do yo create this helpers here? I do not really liked because: 

- In case of a fail, it'll report in the helper. 
- Is not a helper due is making a assert on that. 


For me this function should be a function into runtime/benchmark.go

jrfastab replied:
I'm just going to remove it for now if the netperf fails we can get that from the error there.

eloycoto commented on test/helpers/docker.go:
```diff
@@ -92,6 +92,23 @@ func (s *SSHMeta) ContainerInspectNet(name string) (map[string]string, error) {
 
+// ContainerInspectIP returns the IPv4 and IPv6 strings for the container with
+// the provided name
+func (s *SSHMeta) ContainerInspectIP(name string) (string, string, error) {
```
Why you added a new helper? This is already in place with `vm.ContainerInspectNet()` no?

jrfastab replied:
Just a bit easier to use IMO returns a tuple vs a map but I have no problem pushing it over to InspectNet.

jrfastab replied:
Although looking at it now I only use it in a couple spots so really doesn't make much sense. Using InspectNet() now.

eloycoto commented on test/runtime/benchmark.go:
```diff
@@ -0,0 +1,194 @@
+		}, 300)
+	})
+
+	Context("Benchmark Netperf Tests Sockops-Enabled", func() {
```
This context does not have test on it.

eloycoto commented on test/helpers/benchmark.go:
```diff
@@ -0,0 +1,57 @@
+	"github.com/cilium/cilium/test/ginkgo-ext"
+)
+
+func (s *SSHMeta) superNetperfRRIPv4(client string, server string, num int) *CmdRes {
```
could we move this functions to the benchmark code? Is not really a global helper at all and has some asserts  in there.

jrfastab replied:
sure. At some point I also would like K8s netperf tests but we can uplevel then.

eloycoto commented on test/runtime/benchmark.go:
```diff
@@ -0,0 +1,259 @@
+		vm.ContainerCreate(helpers.Server, helpers.NetperfImage, helpers.CiliumDockerNetwork, "-l id.server")
+		vm.PolicyDelAll()
+		vm.WaitEndpointsReady()
+		err := helpers.WithTimeout(func() bool {
```
This is not needed at all, EndpointsReady will take care of that

jrfastab replied:
sure will remove `err:= helpers.WithTimeout(func() bool { ... `

eloycoto commented on test/runtime/benchmark.go:
```diff
@@ -0,0 +1,259 @@
+		By("create Server containers")
+		vm.ContainerCreate(helpers.Server, helpers.NetperfImage, helpers.CiliumDockerNetwork, "-l id.server")
+		vm.PolicyDelAll()
+		vm.WaitEndpointsReady()
```
`Expect( 		vm.WaitEndpointsReady() ).To(BeNil(), "Endpoints are not ready")`

jrfastab replied:
updated to suggested Expect(...)

eloycoto reviewed (changes requested):
The only thing that worried me is the `helpers/benchmark.go` to test helpers, due is not a real helper. 

Thanks for the work :-)

jrfastab committed:
cilium: add benchmark test class with base netperf test

Add a new class of tests for running benchmarks on the datapath. This
allows us to repeat performance test easily.

At first some basic netperf tests are created but these can be easily
extended. The output is reported in a perf.log file in the normal
test_results directory.

What about kubernetes-netperftest.sh? The kubernetes script is great
but gives us little control over how the netperf tests are executed
and what configurations are used from Cilium side. By writing our
own (and it doesn't take that much work) we get native ginkgo support
and full control which we can leverage for further testing.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

eloycoto reviewed (approved):
:clap: :clap: :clap: :clap:
